src/foam/core/AbstractAgency.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'AbstractAgency',
+  abstract: true,
+  implements: [ 'foam.core.Agency' ],
+
+  javaImports: [
+    'foam.nanos.logger.Loggers',
+    'java.util.Timer',
+    'java.util.TimerTask',
+    'java.util.concurrent.ConcurrentHashMap'
+  ],
+
+  javaCode: `
+    protected static final Timer TIMER = new Timer();
+    protected static final ConcurrentHashMap<String, TimerTask> TASK_QUEUE = new ConcurrentHashMap<>();
+
+    public void schedule(X x, ContextAgent agent, String key, long delay) {
+      if ( delay <= 0 ) {
+        submit(x, agent, key);
+        return;
+      }
+
+      var task = TASK_QUEUE.get(key);
+
+      // Cancel the task if it's not yet executed and being re-scheduled
+      if ( task != null
+        && task.scheduledExecutionTime() - System.currentTimeMillis() > 0
+      ) {
+        task.cancel();
+        Loggers.logger(x, this).debug("Merge scheduled task", key);
+      }
+
+      // Schedule executing the agent
+      task = new TimerTask() {
+        @Override
+        public void run() {
+          agent.execute(x);
```

Comment:
you'll probably need to wrap this in a `try {} finally {}`

Reply:
Fixed in https://github.com/kgrgreer/foam3/pull/1027/commits/274dbb839bdefb6c777a5c6e09e1e237ee661ea3.